src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -0,0 +1,329 @@
+        /// </summary>
+        public string Name { get; }
+
+        public bool Equals(TypeName? other)
+            => other is not null
+            && other._rankOrModifier == _rankOrModifier
+            // try to prevent from allocations if possible (AssemblyQualifiedName can allocate)
+            && ((other._assemblyName is null && _assemblyName is null)
+             || (other._assemblyName is not null && _assemblyName is not null))
+            && other.AssemblyQualifiedName == AssemblyQualifiedName;
+
+        public override bool Equals(object? obj) => Equals(obj as TypeName);
+
+        public override int GetHashCode() => AssemblyQualifiedName.GetHashCode();
+
+        /// <summary>
+        /// Represents the total number of <see cref="TypeName"/> instances that are used to describe
+        /// this instance, including any generic arguments or underlying types.
+        /// </summary>
+        /// <remarks>
+        /// <para>This value is computed every time this method gets called, it's not cached.</para>
+        /// <para>There's not really a parallel concept to this in reflection. Think of it
+        /// as the total number of <see cref="TypeName"/> instances that would be created if
+        /// you were to totally deconstruct this instance and visit each intermediate <see cref="TypeName"/>
+        /// that occurs as part of deconstruction.</para>
+        /// <para>"int" and "Person" each have complexities of 1 because they're standalone types.</para>
+        /// <para>"int[]" has a node count of 2 because to fully inspect it involves inspecting the
+        /// array type itself, <em>plus</em> unwrapping the underlying type ("int") and inspecting that.</para>
+        /// <para>
+        /// "Dictionary&lt;string, List&lt;int[][]&gt;&gt;" has node count 8 because fully visiting it
+        /// involves inspecting 8 <see cref="TypeName"/> instances total:
+        /// <list type="bullet">
+        /// <item>Dictionary&lt;string, List&lt;int[][]&gt;&gt; (the original type)</item>
+        /// <item>Dictionary`2 (the generic type definition)</item>
+        /// <item>string (a type argument of Dictionary)</item>
+        /// <item>List&lt;int[][]&gt; (a type argument of Dictionary)</item>
+        /// <item>List`1 (the generic type definition)</item>
+        /// <item>int[][] (a type argument of List)</item>
+        /// <item>int[] (the underlying type of int[][])</item>
+        /// <item>int (the underlying type of int[])</item>
+        /// </list>
+        /// </para>
+        /// </remarks>
+        public int GetNodeCount()
+        {
+            int result = 1;
+
+            if (IsNested)
+            {
+                result = checked(result + DeclaringType.GetNodeCount());
+            }
+            else if (IsConstructedGenericType)
+            {
+                result = checked(result + 1);
```

Comment:
I assume you wrote it like this rather than `result = 2;` to help with possible future refactoring?